server/s3_auth_client.cc

```diff
@@ -675,11 +675,19 @@ bool S3AuthClient::setup_auth_request_body() {
   const char *full_path = request->c_get_full_encoded_path();
   if (full_path != NULL) {
     std::string uri_full_path = full_path;
-    // in encoded uri path space is encoding as '+'
-    // but cannonical request should have '%20' for verification.
-    // decode plus into space, this special handling
-    // is not required for other charcters.
-    S3CommonUtilities::find_and_replaceall(uri_full_path, "+", "%20");
+    // Handle encoding of common special characters in uri path space,
+    // if not then cannonical request generation might break with
+    // SignatureDoesNotMatch error.
+    const char *arr_syms[13] = {"+", "!", "*", "$", "'", "?", "=",
+                                "@", ":", ";", "#", "&", ","};
+    const char *arr_rep[13] = {"%20", "%21", "%2A", "%24", "%27", "%3F", "%3D",
+                               "%40", "%3A", "%3B", "%23", "%26", "%2C"};
+    for (int i = 0; i < 13; i++) {
```

Comment:
lets revise this code, we can avoid calling find replace multiple times

Reply:
@bkirunge7  this check can increase a request time bit...as every request will go through this.

Reply:
done..with latest approach we will be doing only 1 scan and then are replacing encoded chars in uri